graypy/handler.py

```diff
@@ -344,10 +346,141 @@ def _object_to_json(obj):
+
+    def chunk_message(self, message):
+        """Chunk a GELF message
+
+        :param message: GELF message to chunk.
+        :type; bytes
+
+        :return: Iterator of the chunks of a GELF message.
+        :rtype: Iterator[bytes], None
+        """
+        if self.get_message_chunk_number(message) > GELF_MAX_CHUNK_NUMBER:
+            # silently drop messages that exceed the max gelf chunk size
+            return
+        for chunk in self._gen_gelf_chunks(message):
+            yield chunk
+
+
+class GELFWarningChunker(BaseGELFChunker):
+    """GELF UDP message chunker that warns and drops overflowing messages"""
+
+    def chunk_message(self, message):
+        if self.get_message_chunk_number(message) > GELF_MAX_CHUNK_NUMBER:
+            warnings.warn("GELF chunk overflow for message: {}".format(message), GELFChunkOverflowWarning)
+            return
+        for chunk in self._gen_gelf_chunks(message):
+            yield chunk
+
+
+class GELFTruncatingChunker(BaseGELFChunker):
+    """GELF UDP message chunker that truncates overflowing messages"""
+
+    def __init__(self, chunk_size=WAN_CHUNK, gelf_packer=BaseGELFHandler._pack_gelf_dict):
+        """Initialize the GELFTruncatingChunker.
+
+        :param gelf_packer: Function handle for pracking a GELF dictionary
+            into bytes. Should be of the form ``gelf_packer(gelf_dict)``.
+        :type gelf_packer: Callable[dict]
+        """
+        BaseGELFChunker.__init__(self, chunk_size)
+        self.gelf_packer = gelf_packer
+
+    def gen_chunk_overflow_gelf_log(self, raw_message):
+        """Attempt to truncate a chunk overflowing GELF message
+
+        :param raw_message: Original bytes of a chunk overflowing GELF message.
+        :type raw_message: bytes
+
+        :return: Truncated and simplified version of raw_message.
+        :rtype: bytes
+        """
+        try:
+            message = zlib.decompress(raw_message)
+            compressed = True
+        except zlib.error:
+            message = raw_message
+            compressed = False
+
+        gelf_message = json.loads(message.decode("UTF-8"))
+        short_message = gelf_message['short_message']
+
+        while True:
+            gelf_dict = {
+                'version': gelf_message['version'],
+                'host': gelf_message['host'],
+                'short_message': short_message,
+                'timestamp': gelf_message['timestamp'],
+                'level': SYSLOG_LEVELS.get(logging.ERROR, logging.ERROR),
+                'facility': gelf_message['facility'],
+                '_chunk_overflow': True,
+            }
+            packed_message = self.gelf_packer(gelf_dict)
+            if compressed:
+                packed_message = zlib.compress(packed_message)
+            if self.get_message_chunk_number(packed_message) <= GELF_MAX_CHUNK_NUMBER:
+                return packed_message
+            else:
+                short_message = short_message[:-1]
```

Comment:
Would this search be too slow for large messages? I'm thinking we can cap it by setting `short_message` to `gelf_message['short_message'][:self.chunk_size]`.

Reply:
Do you mean to decrement by chunk size? (i.e. `short_message = gelf_message['short_message'][:-self.chunk_size]`) or to simply cap the message length to just one chunk? (and skipping iterating the message size down).

Reply:
Sorry, I mean set `short_message = gelf_messages['short_message'][:-self.chunk_size * GELF_MAX_CHUNK_NUMBER]` and then iterate down. This limit would lower the number of iterations on huge messages. But, on second thought, the limit won't work when the message is compressed. The compression allows us to send larger messages if their compression rate is high.

I wonder if it's too crazy to do something like a binary search for the maximum message size.

The concern I'm trying to voice is that systems tend to log more when they degrade under heavy load. If logging large messages is expensive, it will only amplify the system's load and lead to cascading failures.

Reply:
I've added a shortcut for truncating following the suggestions you posted above. Looking at my past tests calls stacks before and after this short cut shows improvement in the amount of calls done!